README.md
# GNU-RNG-tausx                  

GNU Scientific Library - new random number generator, tausx
Tausx is an extended version of taus2 with a Bays-Durham shuffle.

This repository shows how to add tausx to the GNU Scientific Library.

See the etaus repository for more information about the extended
taus algorithm with a Bays-Durham shuffle.  The test programs in
the etaus repository also apply to the tausx generator, here.

This repository contains difference files for implementing the extended
taus algorithm in the random number generator section of the GNU
Scientific Library.  A new source file, tausx.c, is added to the
rng directory in the GNU Scientific Library.  See tausx.c in this
repository.  The difference file for tausx.c vs taus.c is called
ext.diff.  Tausx.c is on the left (< sign), and taus.c is on
the right (> sign).

Four other files require the following changes in order to compile
tausx.c into the GNU Scientific Library.  The updated code is on the
left.  The original code is on the right.

---------------------------------------------

* Makefile.am - add the name of tausx.c to the list or source programs.

7c7

< libgslrng_la_SOURCES = borosh13.c cmrg.c coveyou.c default.c file.c fishman18.c fishman20.c fishman2x.c gfsr4.c knuthran2.c knuthran.c knuthran2002.c lecuyer21.c minstd.c mrg.c mt.c r250.c ran0.c ran1.c ran2.c ran3.c rand48.c rand.c random.c randu.c ranf.c ranlux.c ranlxd.c ranlxs.c ranmar.c rng.c slatec.c taus.c taus113.c tausx.c transputer.c tt.c types.c uni32.c uni.c vax.c waterman14.c zuf.c inline.c

---

> libgslrng_la_SOURCES = borosh13.c cmrg.c coveyou.c default.c file.c fishman18.c fishman20.c fishman2x.c gfsr4.c knuthran2.c knuthran.c knuthran2002.c lecuyer21.c minstd.c mrg.c mt.c r250.c ran0.c ran1.c ran2.c ran3.c rand48.c rand.c random.c randu.c ranf.c ranlux.c ranlxd.c ranlxs.c ranmar.c rng.c slatec.c taus.c taus113.c transputer.c tt.c types.c uni32.c uni.c vax.c waterman14.c zuf.c inline.c

---------------------------------------------

* Makefile.in - add the name of tausx.lo to the list of object
  modules.

104,105c104,105

< 	taus113.lo tausx.lo transputer.lo tt.lo types.lo uni32.lo \
< 	uni.lo vax.lo waterman14.lo zuf.lo inline.lo

---

> 	taus113.lo transputer.lo tt.lo types.lo uni32.lo uni.lo vax.lo \
> 	waterman14.lo zuf.lo inline.lo

---------------------------------------------

* gsl_rng.h - add the RNG name, gsl_rng_tausx, to the list of
  of generators in the header file for random number generators.

116d115

< GSL_VAR const gsl_rng_type *gsl_rng_tausx;

---------------------------------------------

* types.c - add "tausx" to the list of random number generators.

89d88

<   ADD(gsl_rng_tausx);

---------------------------------------------

Installation instructions.

The latest version of the GNU Scientific Library is version
2.6.  The source code for the GNU Scientific Library may be
downloaded at the following address.

ftp://ftp.gnu.org/gnu/gsl/

ftp://ftp.gnu.org/gnu/gsl/gsl-2.6.tar.gz

README and INSTALL in the main directory explain how to compile
the GNU Scientific Library.

The compile sequence for the GNU Scientific Library is

./configure

make

sudo make install

The static library, libgsl.a, and the shared library, libgsl.so,
are placed in the .libs directory after the compile has completed.

unzip GSL-rng-tausx-master.zip

change directory to GSL-rng-tausx-master

chmod 700 *.sh

ln -s [path]/gsl-2.6 ./gsl-2.6

./inst.sh

If you receive a compile error from inst.sh, check speed.mak to
make sure that the gsl_rng.h header file is pointing to the correct
path in CFLAGS.

Run the speed test.

Make sure that the LD_LIBRARY_PATH variable in runspeed.sh is
pointing to the correct address path.

./runspeed.sh GSL_rng_name

Examples:

./runspeed.sh tausx

./runspeed.sh taus2

To de-install the speed test,

./deinst.sh

---------------------------------------------

Speed test program.

Program speed.c in this repository runs a speed test on a
GNU Scientific Library random number generator.  The program
calculates the number of ticks spent performing 100 million
get operations.

Tausx runs about 15% slower than taus2.



ChangeLog
2020-02-20  aquila57 at github.com

	* added tausx.c
	* changed types.c to include tausx
	* changed Makefile.am to include tausx.c
	* changed Makefile.in to include tausx.lo
	* changed gsl/gsl_rng.h to include gsl_rng_tausx
	tausx is taus2 with a Bays-Durham shuffle.

2012-09-10  Rhys Ulerich <rhys.ulerich@gmail.com>

	* ranf.c Add include for gsl_sys.h to fix MSVC build.
	Thanks to Brian Gladman for the suggestion.

2011-04-28  Brian Gough  <bjg@network-theory.co.uk>

	* ranlxs.c (ranlxs_set_lux): enforce maximum seed of
	2^31-1 (0x7FFFFFFF) for ranlxs

2010-01-07  Brian Gough  <bjg@network-theory.co.uk>

	* ran3.c (ran3_set): added a check for negative j

2009-07-09  Brian Gough  <bjg@network-theory.co.uk>

	* rng.c (gsl_rng_free): handle NULL argument in free

2008-10-13  Brian Gough  <bjg@network-theory.co.uk>

	* file.c: added (char *) to allow compilation with C++ compiler

2008-07-03  Brian Gough  <bjg@hp2.network-theory.co.uk>

	* rng.c: moved inline functions to inline.c

	* gsl_rng.h: use new inline declarations

	* inline.c: handle inline functions separately

	* Makefile.am (INCLUDES): use top_srcdir instead of top_builddir

2007-01-28  Brian Gough  <bjg@network-theory.co.uk>

	* knuthran2002.c: added revised version from 9th printing

2007-01-04  Brian Gough  <bjg@network-theory.co.uk>

	* default.c (gsl_rng_env_setup): send newline to stderr not stdout

2006-04-13  Brian Gough  <bjg@network-theory.co.uk>

	* default.c (gsl_rng_env_setup): print a newline after list of
	generators in error message

2006-03-16  Brian Gough  <bjg@network-theory.co.uk>

	* test.c (main): added taus2 test

2006-02-19  Brian Gough  <bjg@network-theory.co.uk>

	* rng.c: added note about why range=max-min not max-min+1

2005-12-16  Brian Gough  <bjg@network-theory.co.uk>

	* rng.c (gsl_rng_uniform_int): catch the case n = 0 and return an
	error

2003-07-23  Brian Gough  <bjg@network-theory.co.uk>

	* file.c: added fwrite/fread functions

2003-06-12  Brian Gough  <bjg@network-theory.co.uk>

	* Makefile.am: removed benchmark programs from default build

2003-06-02  Brian Gough  <bjg@network-theory.co.uk>

	* waterman14.c: increased RAND_MIN to 1

	* lecuyer21.c: corrections to RAND_MIN, RAND_MAX, floating point
	denominator and seeding modulus

	* fishman20.c (ran_get): use schrage multiplication to avoid
	overflow

	* coveyou.c: corrected value of RAND_MIN to 1 and RAND_MAX to 2^32-1

	* borosh13.c: increased RAND_MIN to 1

	* knuthran2.c (ran_get): use schrage multiplication to avoid
	overflow in intermediate results

	* fishman2x.c (ran_get): use schrage multiplication to avoid
	overflow in intermediate results

	* fishman18.c (ran_get): use schrage multiplication to avoid
	overflow in intermediate results

	* schrage.c (schrage): utility functions for multiplication of
	long integers

	* test.c (main): updated incorrect test values for
	gsl_rng_fishman18 gsl_rng_fishman2x gsl_rng_knuthran2

Mon Nov 25 19:27:10 2002  Brian Gough  <bjg@network-theory.co.uk>

	* taus.c (taus2_set): fixed bug in seeding for s3 (test should be
 	for s3<16, not s3<8)

Sun Nov  3 14:40:43 2002  Brian Gough  <bjg@network-theory.co.uk>

	* taus.c (taus2_set): fixed bug in seeding for s2 < 8

Mon Jun 17 21:04:10 2002  Brian Gough  <bjg@network-theory.co.uk>

	* mt.c (mt_set): ensure state counter mti is initialised by
 	seeding routine

Wed May 29 21:52:11 2002  Brian Gough  <bjg@network-theory.co.uk>

	* randu.c: removed unused variable m

Tue May  7 22:34:52 2002  Brian Gough  <bjg@network-theory.co.uk>

	* mt.c (mt_1999_set): updated seeding procedure according to new
 	release of MT19937 from the original authors.

Fri Apr 26 21:17:05 2002  Brian Gough  <bjg@network-theory.co.uk>

	* taus.c (taus2_set): added alternate seeding procedure as
 	described in erratum to P.L'Ecuyer's paper.

Wed Apr 17 19:37:49 2002  Brian Gough  <bjg@network-theory.co.uk>

	* test.c (main): added missing test for gsl_rng_mt19937_1998

	* gsl_rng.h: added missing declaration for gsl_rng_mt19937_1998

Sun Dec  2 15:45:24 2001  Brian Gough  <bjg@network-theory.co.uk>

	* Added new generators borosh13, coveyou, fishman18, fishman20,
 	fishman2x, knuthran, knuthran2, lecuyer21, waterman14 from Knuth's
 	Seminumerical Algorithms 3rd Ed.  Implemented by Carlo Perassi.

	* gfsr4.c (gfsr4_get_double): increased divisor for double to
 	2^32, avoids generating exact result of 1.0 as specified in the
 	documentation

Mon Sep  3 10:32:01 2001  Brian Gough  <bjg@network-theory.co.uk>

	* mt.c (mt_1998_set): renamed macro to avoid duplicate definition

Fri Aug 31 17:49:37 2001  Brian Gough  <bjg@network-theory.co.uk>

	* mt.c (mt_1998_set): added the original (buggy) MT19937 seeding
 	routine as mt19937_1998 for compatibility.

Wed May  2 15:35:38 2001  Brian Gough  <bjg@network-theory.co.uk>

	* ran1.c (ran1_get_double): use float constants for comparison
	for compatibility with original Numerical Recipes routines

	* ran2.c (ran2_get_double): ditto
	
Fri Apr 27 18:47:07 2001  Brian Gough  <bjg@network-theory.co.uk>

	* types.c (gsl_rng_types_setup): added void to make prototype
 	valid in ansi c

Mon Apr 16 20:03:07 2001  Brian Gough  <bjg@network-theory.co.uk>

	* default.c (gsl_rng_env_setup): removed spurious argument to
 	fprintf

Tue Jan 23 13:24:26 2001  Brian Gough  <bjg@network-theory.co.uk>

	* types.c (gsl_rng_types_setup): provide a function that returns a
 	list of all the generator types

	* default.c (gsl_rng_env_setup): get the list of generators from a
 	function rather than having a list in the code itself.  Display a
 	list of the valid generators if the user provides an incorrect one.

Fri Dec  8 20:30:58 2000  Brian Gough  <bjg@network-theory.co.uk>

	* ranlxs.c: renamed internal function ranlxs_set_impl to
 	ranlxs_set
	
	* ranlxd.c: renamed internal function ranlxd_set_impl to
 	ranlxd_set

	* ranlux.c: renamed internal function ranlux_set_impl to
 	ranlux_set

	* random.c: renamed internal function random_get_impl function to

speed.c
/* speed.c - Speed test for GSL generators  Version 0.1.0            */
/* Copyright (C) 2020 aquila57 at github.com                         */

/* This program is free software; you can redistribute it and/or     */
/* modify it under the terms of the GNU General Public License as    */
/* published by the Free Software Foundation; either version 2 of    */
/* the License, or (at your option) any later version.               */

/* This program is distributed in the hope that it will be useful,   */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of    */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the      */
/* GNU General Public License for more details.                      */

/* You should have received a copy of the GNU General Public License */
/* along with this program; if not, write to:                        */

   /* Free Software Foundation, Inc.                                 */
   /* 59 Temple Place - Suite 330                                    */
   /* Boston, MA 02111-1307, USA.                                    */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <sys/times.h>
/* this is the new version of gsl_rng.h */
#include <gsl_rng.h>

#define SMPLS 100000000

int main(void)
   {
   int i;                      /* loop counter */
   unsigned long int x;        /* result of get */
   clock_t begclk;             /* current number of ticks */
   clock_t endclk;             /* current number of ticks */
   int duration;               /* duration of test in ticks */
   struct tms t;               /* structure used by times() */
   gsl_rng *rng;               /* RNG structure (state) */
   gsl_rng_type *typ;          /* RNG name */
   /************************************************/
   /* initialize the GSL random number generator   */
   /************************************************/
   gsl_rng_env_setup();
   typ = (gsl_rng_type *) gsl_rng_default;
   rng = (gsl_rng *) gsl_rng_alloc(typ);
   /* set the initial seed to 1 [the default] */
   gsl_rng_set(rng,1);
   /* print column headings */
   printf("Generator       #Ticks\n");
   /* get #clock ticks since boot */
   begclk = times(&t);         /* starting time in ticks */
   i = SMPLS;                  /* number of samples */
   while (i--)
      {
      x = gsl_rng_get(rng);
      } /* for each iteration of get */
   endclk = times(&t);         /* ending time in ticks */
   duration = endclk - begclk;   /* calculate duration */
   printf("%s %15d\n", gsl_rng_name(rng), duration);
   gsl_rng_free(rng);    /* de-allocate the RNG */
   x = x;                /* to avoid a warning message */
   return(0);
   } /* main */


types.c
/* rng/types.c
 * 
 * Copyright (C) 2001, 2007 Brian Gough
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

// #include <config.h>
#include <stdlib.h>
// #include <gsl/gsl_rng.h>
#include "gsl_rng.h"

#define N 100

const gsl_rng_type * gsl_rng_generator_types[N];

#define ADD(t) {if (i==N) abort(); gsl_rng_generator_types[i] = (t); i++; };

const gsl_rng_type **
gsl_rng_types_setup (void)
{
  int i = 0;

  ADD(gsl_rng_borosh13);
  ADD(gsl_rng_cmrg);
  ADD(gsl_rng_coveyou);
  ADD(gsl_rng_fishman18);
  ADD(gsl_rng_fishman20);
  ADD(gsl_rng_fishman2x);
  ADD(gsl_rng_gfsr4);
  ADD(gsl_rng_knuthran);
  ADD(gsl_rng_knuthran2);
  ADD(gsl_rng_knuthran2002);
  ADD(gsl_rng_lecuyer21);
  ADD(gsl_rng_minstd);
  ADD(gsl_rng_mrg);
  ADD(gsl_rng_mt19937);
  ADD(gsl_rng_mt19937_1999);
  ADD(gsl_rng_mt19937_1998);
  ADD(gsl_rng_r250);
  ADD(gsl_rng_ran0);
  ADD(gsl_rng_ran1);
  ADD(gsl_rng_ran2);
  ADD(gsl_rng_ran3);
  ADD(gsl_rng_rand);
  ADD(gsl_rng_rand48);
  ADD(gsl_rng_random128_bsd);
  ADD(gsl_rng_random128_glibc2);
  ADD(gsl_rng_random128_libc5);
  ADD(gsl_rng_random256_bsd);
  ADD(gsl_rng_random256_glibc2);
  ADD(gsl_rng_random256_libc5);
  ADD(gsl_rng_random32_bsd);
  ADD(gsl_rng_random32_glibc2);
  ADD(gsl_rng_random32_libc5);
  ADD(gsl_rng_random64_bsd);
  ADD(gsl_rng_random64_glibc2);
  ADD(gsl_rng_random64_libc5);
  ADD(gsl_rng_random8_bsd);
  ADD(gsl_rng_random8_glibc2);
  ADD(gsl_rng_random8_libc5);
  ADD(gsl_rng_random_bsd);
  ADD(gsl_rng_random_glibc2);
  ADD(gsl_rng_random_libc5);
  ADD(gsl_rng_randu);
  ADD(gsl_rng_ranf);
  ADD(gsl_rng_ranlux);
  ADD(gsl_rng_ranlux389);
  ADD(gsl_rng_ranlxd1);
  ADD(gsl_rng_ranlxd2);
  ADD(gsl_rng_ranlxs0);
  ADD(gsl_rng_ranlxs1);
  ADD(gsl_rng_ranlxs2);
  ADD(gsl_rng_ranmar);
  ADD(gsl_rng_slatec);
  ADD(gsl_rng_taus);
  ADD(gsl_rng_taus2);
  ADD(gsl_rng_tausx);
  ADD(gsl_rng_taus113);
  ADD(gsl_rng_transputer);
  ADD(gsl_rng_tt800);
  ADD(gsl_rng_uni);
  ADD(gsl_rng_uni32);
  ADD(gsl_rng_vax);
  ADD(gsl_rng_waterman14);
  ADD(gsl_rng_zuf);
  ADD(0);

  return gsl_rng_generator_types;
}



gsl_rng.h
/* rng/gsl_rng.h
 * 
 * Copyright (C) 1996, 1997, 1998, 1999, 2000, 2004, 2007 James Theiler, Brian Gough
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#ifndef __GSL_RNG_H__
#define __GSL_RNG_H__
#include <stdlib.h>
#include <gsl/gsl_types.h>
#include <gsl/gsl_errno.h>
#include <gsl/gsl_inline.h>

#undef __BEGIN_DECLS
#undef __END_DECLS
#ifdef __cplusplus
# define __BEGIN_DECLS extern "C" {
# define __END_DECLS }
#else
# define __BEGIN_DECLS /* empty */
# define __END_DECLS /* empty */
#endif

__BEGIN_DECLS

typedef struct
  {
    const char *name;
    unsigned long int max;
    unsigned long int min;
    size_t size;
    void (*set) (void *state, unsigned long int seed);
    unsigned long int (*get) (void *state);
    double (*get_double) (void *state);
  }
gsl_rng_type;

typedef struct
  {
    const gsl_rng_type * type;
    void *state;
  }
gsl_rng;


/* These structs also need to appear in default.c so you can select
   them via the environment variable GSL_RNG_TYPE */

GSL_VAR const gsl_rng_type *gsl_rng_borosh13;
GSL_VAR const gsl_rng_type *gsl_rng_coveyou;
GSL_VAR const gsl_rng_type *gsl_rng_cmrg;
GSL_VAR const gsl_rng_type *gsl_rng_fishman18;
GSL_VAR const gsl_rng_type *gsl_rng_fishman20;
GSL_VAR const gsl_rng_type *gsl_rng_fishman2x;
GSL_VAR const gsl_rng_type *gsl_rng_gfsr4;
GSL_VAR const gsl_rng_type *gsl_rng_knuthran;
GSL_VAR const gsl_rng_type *gsl_rng_knuthran2;
GSL_VAR const gsl_rng_type *gsl_rng_knuthran2002;
GSL_VAR const gsl_rng_type *gsl_rng_lecuyer21;
GSL_VAR const gsl_rng_type *gsl_rng_minstd;
GSL_VAR const gsl_rng_type *gsl_rng_mrg;
GSL_VAR const gsl_rng_type *gsl_rng_mt19937;
GSL_VAR const gsl_rng_type *gsl_rng_mt19937_1999;
GSL_VAR const gsl_rng_type *gsl_rng_mt19937_1998;
GSL_VAR const gsl_rng_type *gsl_rng_r250;
GSL_VAR const gsl_rng_type *gsl_rng_ran0;
GSL_VAR const gsl_rng_type *gsl_rng_ran1;
GSL_VAR const gsl_rng_type *gsl_rng_ran2;
GSL_VAR const gsl_rng_type *gsl_rng_ran3;
GSL_VAR const gsl_rng_type *gsl_rng_rand;
GSL_VAR const gsl_rng_type *gsl_rng_rand48;
GSL_VAR const gsl_rng_type *gsl_rng_random128_bsd;
GSL_VAR const gsl_rng_type *gsl_rng_random128_glibc2;
GSL_VAR const gsl_rng_type *gsl_rng_random128_libc5;
GSL_VAR const gsl_rng_type *gsl_rng_random256_bsd;
GSL_VAR const gsl_rng_type *gsl_rng_random256_glibc2;
GSL_VAR const gsl_rng_type *gsl_rng_random256_libc5;
GSL_VAR const gsl_rng_type *gsl_rng_random32_bsd;
GSL_VAR const gsl_rng_type *gsl_rng_random32_glibc2;
GSL_VAR const gsl_rng_type *gsl_rng_random32_libc5;
GSL_VAR const gsl_rng_type *gsl_rng_random64_bsd;
GSL_VAR const gsl_rng_type *gsl_rng_random64_glibc2;
GSL_VAR const gsl_rng_type *gsl_rng_random64_libc5;
GSL_VAR const gsl_rng_type *gsl_rng_random8_bsd;
GSL_VAR const gsl_rng_type *gsl_rng_random8_glibc2;
GSL_VAR const gsl_rng_type *gsl_rng_random8_libc5;
GSL_VAR const gsl_rng_type *gsl_rng_random_bsd;
GSL_VAR const gsl_rng_type *gsl_rng_random_glibc2;
GSL_VAR const gsl_rng_type *gsl_rng_random_libc5;
GSL_VAR const gsl_rng_type *gsl_rng_randu;
GSL_VAR const gsl_rng_type *gsl_rng_ranf;
GSL_VAR const gsl_rng_type *gsl_rng_ranlux;
GSL_VAR const gsl_rng_type *gsl_rng_ranlux389;
GSL_VAR const gsl_rng_type *gsl_rng_ranlxd1;
GSL_VAR const gsl_rng_type *gsl_rng_ranlxd2;
GSL_VAR const gsl_rng_type *gsl_rng_ranlxs0;
GSL_VAR const gsl_rng_type *gsl_rng_ranlxs1;
GSL_VAR const gsl_rng_type *gsl_rng_ranlxs2;
GSL_VAR const gsl_rng_type *gsl_rng_ranmar;
GSL_VAR const gsl_rng_type *gsl_rng_slatec;
GSL_VAR const gsl_rng_type *gsl_rng_taus;
GSL_VAR const gsl_rng_type *gsl_rng_taus2;
GSL_VAR const gsl_rng_type *gsl_rng_tausx;
GSL_VAR const gsl_rng_type *gsl_rng_taus113;
GSL_VAR const gsl_rng_type *gsl_rng_transputer;
GSL_VAR const gsl_rng_type *gsl_rng_tt800;
GSL_VAR const gsl_rng_type *gsl_rng_uni;
GSL_VAR const gsl_rng_type *gsl_rng_uni32;
GSL_VAR const gsl_rng_type *gsl_rng_vax;
GSL_VAR const gsl_rng_type *gsl_rng_waterman14;
GSL_VAR const gsl_rng_type *gsl_rng_zuf;

const gsl_rng_type ** gsl_rng_types_setup(void);

GSL_VAR const gsl_rng_type *gsl_rng_default;
GSL_VAR unsigned long int gsl_rng_default_seed;

gsl_rng *gsl_rng_alloc (const gsl_rng_type * T);
int gsl_rng_memcpy (gsl_rng * dest, const gsl_rng * src);
gsl_rng *gsl_rng_clone (const gsl_rng * r);

void gsl_rng_free (gsl_rng * r);

void gsl_rng_set (const gsl_rng * r, unsigned long int seed);
unsigned long int gsl_rng_max (const gsl_rng * r);
unsigned long int gsl_rng_min (const gsl_rng * r);
const char *gsl_rng_name (const gsl_rng * r);

int gsl_rng_fread (FILE * stream, gsl_rng * r);
int gsl_rng_fwrite (FILE * stream, const gsl_rng * r);

size_t gsl_rng_size (const gsl_rng * r);
void * gsl_rng_state (const gsl_rng * r);

void gsl_rng_print_state (const gsl_rng * r);

const gsl_rng_type * gsl_rng_env_setup (void);

INLINE_DECL unsigned long int gsl_rng_get (const gsl_rng * r);
INLINE_DECL double gsl_rng_uniform (const gsl_rng * r);
INLINE_DECL double gsl_rng_uniform_pos (const gsl_rng * r);
INLINE_DECL unsigned long int gsl_rng_uniform_int (const gsl_rng * r, unsigned long int n);

#ifdef HAVE_INLINE

INLINE_FUN unsigned long int
gsl_rng_get (const gsl_rng * r)
{
  return (r->type->get) (r->state);
}

INLINE_FUN double
gsl_rng_uniform (const gsl_rng * r)
{
  return (r->type->get_double) (r->state);
}

INLINE_FUN double
gsl_rng_uniform_pos (const gsl_rng * r)
{
  double x ;
  do
    {
      x = (r->type->get_double) (r->state) ;
    }
  while (x == 0) ;

  return x ;
}

/* Note: to avoid integer overflow in (range+1) we work with scale =
   range/n = (max-min)/n rather than scale=(max-min+1)/n, this reduces
   efficiency slightly but avoids having to check for the out of range
   value.  Note that range is typically O(2^32) so the addition of 1
   is negligible in most usage. */

INLINE_FUN unsigned long int
gsl_rng_uniform_int (const gsl_rng * r, unsigned long int n)
{
  unsigned long int offset = r->type->min;
  unsigned long int range = r->type->max - offset;
  unsigned long int scale;
  unsigned long int k;

  if (n > range || n == 0) 
    {
      GSL_ERROR_VAL ("invalid n, either 0 or exceeds maximum value of generator",
                     GSL_EINVAL, 0) ;
    }

  scale = range / n;

  do
    {
      k = (((r->type->get) (r->state)) - offset) / scale;
    }
  while (k >= n);

  return k;
}
#endif /* HAVE_INLINE */

__END_DECLS

#endif /* __GSL_RNG_H__ */


tausx.c
/* rng/tausx.c
 * 
 * Copyright (C) 2020  aquila57 at github.com
 * 
 * This is an an extension to taus.c with a Bays-Durham shuffle.
 *
 * Copyright for taus.c:
 *
 * Copyright (C) 1996, 1997, 1998, 1999, 2000, 2007 James Theiler, Brian Gough
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <config.h>
#include <stdlib.h>
#include <gsl/gsl_rng.h>

/* This is a maximally equidistributed combined Tausworthe
   generator. The sequence for tausx_get() is,

   x_{n-2} = x_{n-1}
   x_{n-1} = x_n

   x_n = (s1_n ^ s2_n ^ s3_n) 

   s1_{n+1} = (((s1_n & 4294967294) <<12) ^ (((s1_n <<13) ^ s1_n) >>19))
   s2_{n+1} = (((s2_n & 4294967288) << 4) ^ (((s2_n << 2) ^ s2_n) >>25))
   s3_{n+1} = (((s3_n & 4294967280) <<17) ^ (((s3_n << 3) ^ s3_n) >>11))

   index = x_{x-2} >> 18
   temp = sx[index]
   sx[index] = x_n
   x_n = temp

   computed modulo 2^32. In the three formulas above '^' means
   exclusive-or (C-notation), not exponentiation. Note that the
   algorithm relies on the properties of 32-bit unsigned integers (it
   is formally defined on bit-vectors of length 32). I have added a
   bitmask to make it work on 64 bit machines.

   sx is a random list of 16384 32-bit numbers.
   temp is a temporary 32-bit number.
   index is a 32-bit signed integer as a subscript to sx.

   We initialize the generator with s1_1 .. s3_1 = s_n MOD m, where
   s_n = (69069 * s_{n-1}) mod 2^32, and s_0 = s is the user-supplied
   seed.

   After s1, s2, and s3 have been initialized as with taus2, the array sx
   is initialized by performing the Tausworthe algorithm for each element
   in the array.

   The theoretical value of x_{17391} is 1132370745.  The subscript
   17391 means (1) seed the generator with s=1 (2) do six warm-up
   iterations, (3) initialize sx, (4) then do 10000 actual iterations.

   The period of this generator approximates the factorial of 16384,
   or 10^61036.  It is orders of magnitude greater than the period
   of the original taus generator.

   From: P. L'Ecuyer, "Maximally Equidistributed Combined Tausworthe
   Generators", Mathematics of Computation, 65, 213 (1996), 203--213.

   This is available on the net from L'Ecuyer's home page,

   http://www.iro.umontreal.ca/~lecuyer/myftp/papers/tausme.ps
   ftp://ftp.iro.umontreal.ca/pub/simulation/lecuyer/papers/tausme.ps 

   Update: April 2002

   There is an erratum in the paper "Tables of Maximally
   Equidistributed Combined LFSR Generators", Mathematics of
   Computation, 68, 225 (1999), 261--269:
   http://www.iro.umontreal.ca/~lecuyer/myftp/papers/tausme2.ps

        ... the k_j most significant bits of z_j must be non-
        zero, for each j. (Note: this restriction also applies to the 
        computer code given in [4], but was mistakenly not mentioned in
        that paper.)
   
   This affects the seeding procedure by imposing the requirement
   s1 > 1, s2 > 7, s3 > 15.

   The generator taus2 has been added to satisfy this requirement.
   The original taus generator is unchanged.

   Update: November 2002

   There was a bug in the correction to the seeding procedure for s2.
   It affected the following seeds 254679140 1264751179 1519430319
   2274823218 2529502358 3284895257 3539574397 (s2 < 8).

*/

static inline unsigned long int tausx_get (void *vstate);
static double tausx_get_double (void *vstate);
static void tausx_set (void *state, unsigned long int s);

typedef struct
  {
    unsigned long int s1, s2, s3;
    unsigned long int xpp;        /* x_{n-2}           */
    unsigned long int xp;         /* x_{n-1}           */       
    unsigned long int xn;         /* s1^s2^s3          */
    unsigned long int ix;         /* index to sx array */       
    unsigned long int tx;         /* temporary xn      */       
    unsigned long int sx[16384];     /* shuffle array  */
  }
tausx_state_t;

static inline unsigned long tausx_get (void *vstate)
{

  tausx_state_t *state = (tausx_state_t *) vstate;

#define MASK 0xffffffffUL
#define TAUSWORTHE(s,a,b,c,d) (((s &c) <<d) &MASK) ^ ((((s <<a) &MASK)^s) >>b)

  state->xpp = state->xp;
  state->xp  = state->xn;

  state->s1 = TAUSWORTHE (state->s1, 13, 19, 4294967294UL, 12);
  state->s2 = TAUSWORTHE (state->s2, 2, 25, 4294967288UL, 4);
  state->s3 = TAUSWORTHE (state->s3, 3, 11, 4294967280UL, 17);

  state->xn  = (state->s1 ^ state->s2 ^ state->s3);
  state->ix = state->xpp >> 18;
  state->tx = state->sx[state->ix];
  state->sx[state->ix] = state->xn;
  state->xn = state->tx;

  return state->xn;
}

static double tausx_get_double (void *vstate)
{
  return tausx_get (vstate) / 4294967296.0 ;
}

static void tausx_set (void *vstate, unsigned long int s)
{
  tausx_state_t *state = (tausx_state_t *) vstate;

  if (s == 0)
    s = 1;      /* default seed is 1 */

/* The following is the taus2 adjustment to the initial seeds */

#define LCG(n) ((69069 * n) & 0xffffffffUL)
  state->s1 = LCG (s);
  if (state->s1 < 2) state->s1 += 2UL;
  state->s2 = LCG (state->s1);
  if (state->s2 < 8) state->s2 += 8UL;
  state->s3 = LCG (state->s2);
  if (state->s3 < 16) state->s3 += 16UL;

  inline void tausx_get_simple(tausx_state_t *state)
     {
#define MASK 0xffffffffUL
#define TAUSWORTHE(s,a,b,c,d) (((s &c) <<d) &MASK) ^ ((((s <<a) &MASK)^s) >>b)
     state->s1 = TAUSWORTHE (state->s1, 13, 19, 4294967294UL, 12);
     state->s2 = TAUSWORTHE (state->s2, 2, 25, 4294967288UL, 4);
     state->s3 = TAUSWORTHE (state->s3, 3, 11, 4294967280UL, 17);
     } /* tausx_get_simple(state) */

  inline void tausx_init_array(tausx_state_t *state)
     {
     unsigned long int *p,*q;
     p = (unsigned long int *) state->sx;
     q = (unsigned long int *) state->sx + 16384;
     while (p < q)
        {
        tausx_get_simple (state);
	*p++ = (state->s1 ^ state->s2 ^ state->s3);
	} /* for each element in shuffle array */
     } /* tausx_init_array(state) */

  /* "warm it up" */
  tausx_get_simple (state);
  tausx_get_simple (state);
  tausx_get_simple (state);
  tausx_get_simple (state);
  tausx_get_simple (state);
  tausx_get_simple (state);
  /* initialize the Bays-Durham shuffle array  */
  /* to random 32-bit non-zero numbers         */
  tausx_init_array (state);
  return;
}

static const gsl_rng_type tausx_type =
{"tausx",                       /* name */
 0xffffffffUL,                  /* RAND_MAX */
 0,                             /* RAND_MIN */
 sizeof (tausx_state_t),
 &tausx_set,
 &tausx_get,
 &tausx_get_double};

const gsl_rng_type *gsl_rng_tausx = &tausx_type;
